Metrics for monitoring endpoints

51-code opened the pull request:
Fixes #8 

Added metrics for:
- Latency per partition (event's enqueued time - read time)
- Depth per partition (last enqueued event's offset - accepted event's offset)
- Data depth estimation (average message size / amount of messages passed)

Dropwizard setup was inspired by [CFE-35 Router.java](https://github.com/teragrep/cfe_35/blob/main/src/main/java/com/teragrep/cfe_35/router/Router.java)

Needs to be tested, but this is the initial code.

UPDATE: Tests done!

kortemik commented:
@StrongestNumber9 please check this out

StrongestNumber9 commented on src/main/java/com/teragrep/aer_01/EventContextConsumer.java:
```diff
@@ -95,8 +127,50 @@ private String getRealHostName() {
+        int messageLength = eventContext.getEventData().getBody().length;
+        String partitionId = eventContext.getPartitionContext().getPartitionId();
+
+        records++;
```
Is this method thread safe? These increments might cause issues with race conditions when threaded

StrongestNumber9 commented on src/main/java/com/teragrep/aer_01/config/MetricsConfig.java:
```diff
@@ -0,0 +1,68 @@
+
+    public MetricsConfig(Sourceable configSource) {
+        this.configSource = configSource;
+        this.prometheusPort = getPrometheusPort();
```
Superfluous with the getPrometheusPort() command

StrongestNumber9 commented on src/main/java/com/teragrep/aer_01/config/MetricsConfig.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    /**
+     * @return relp.connection.timeout
```
Wrong docs

StrongestNumber9 commented on src/main/java/com/teragrep/aer_01/EventContextConsumer.java:
```diff
@@ -95,8 +127,50 @@ private String getRealHostName() {
+        records++;
+        allSize = allSize + messageLength;
+
+        metricRegistry.gauge(name(EventContextConsumer.class, "latency-seconds", partitionId), () -> new Gauge<Long>() {
+            @Override
+            public Long getValue() {
+                return Instant.now().getEpochSecond() - eventContext.getEventData().getEnqueuedTime().getEpochSecond();
+            }
+        });
+        metricRegistry.gauge(name(EventContextConsumer.class, "depth-bytes", partitionId), () -> new Gauge<Long>() {
+            @Override
+            public Long getValue() {
+                return eventContext.getLastEnqueuedEventProperties().getOffset() - eventContext.getEventData().getOffset();
+            }
+        });
```
I am fairly sure that these gauges should not be recreated on every `accept()` call

51-code replied:
The gauge() function returns the gauge registered under the given name or creates a new one if one doesn't exist. So basically this will create a new gauge when a message is received from a not-seen-before partition, but most of the time it returns the correct gauge for the partition that the message is from.

StrongestNumber9 replied:
Ah getOrCreate method, thats fine

StrongestNumber9 commented on src/main/java/com/teragrep/aer_01/EventContextConsumer.java:
```diff
@@ -46,28 +46,48 @@
+
+    // metrics
+    private int records;
+    private long allSize;
```
Possible race conditions if accepted in multithreaded context

StrongestNumber9 reviewed (changes requested):
Added some notes

51-code commented:
Added testing for the metrics and fixed the issues that came up in the first round of review.

Made a fake object for the `Output` to get rid of dependencies of the `EventContextConsumer` object, and the class was refactored with a few new constructors to allow this fake to be injected. Fakes were also done for multiple objects in Azure library for creating "data" for the object, the `CheckPointlessEventContextFactory` is responsible for data creation.

StrongestNumber9 commented on src/test/java/com/teragrep/aer_01/EventContextConsumerTest.java:
```diff
@@ -0,0 +1,145 @@
+            EventContext eventContext = eventContextFactory.create();
+            eventContextConsumer.accept(eventContext);
+            final double records = 10;
+            for (int i = 0; i < records - 1; i++) {
```
Maybe `for (int i = 1; i <= records; i++)` just for a bit more clarity. 

And theres also an off by one error here if it should send 10 records:

```
jshell> double records = 10;
records ==> 10.0

jshell> for (int i = 0; i < records - 1; i++) {
   ...>     System.out.println(i);
   ...> }
0
1
2
3
4
5
6
7
8
```

StrongestNumber9 commented on src/test/java/com/teragrep/aer_01/EventContextConsumerTest.java:
```diff
@@ -0,0 +1,145 @@
+
+            long depth1 = 0L;
+            final double records = 10;
+            for (int i = 0; i < records - 1; i++) {
```
Same as other note

StrongestNumber9 commented on src/test/java/com/teragrep/aer_01/EventContextConsumerTest.java:
```diff
@@ -0,0 +1,145 @@
+            Gauge<Long> gauge1 = metricRegistry.gauge(name(EventContextConsumer.class, "depth-bytes", "1"));
+            Gauge<Long> gauge2 = metricRegistry.gauge(name(EventContextConsumer.class, "depth-bytes", "2"));
+
+            Assertions.assertEquals(depth1, gauge1.getValue());
```
Assert that depth1 and depth2 are both equal to some well known and expected value as well. What if `getOffset()` just returns `0` every time?

StrongestNumber9 commented on src/test/java/com/teragrep/aer_01/EventContextConsumerTest.java:
```diff
@@ -0,0 +1,145 @@
+        try (EventContextConsumer eventContextConsumer = new EventContextConsumer(configSource, new Output.FakeOutput(), metricRegistry, prometheusPort)) {
+            final double records = 10;
+            long length = 0L;
+            for (int i = 0; i < records; i++) {
```
This seems to be working as expected but change to follow the suit if the others fors are changed

StrongestNumber9 reviewed (commented):
Some minor tweaks

StrongestNumber9 reviewed (approved):
lgtm at this point

kortemik commented on src/main/java/com/teragrep/aer_01/DefaultOutput.java:
```diff
@@ -0,0 +1,189 @@
+                    Thread.sleep(reconnectInterval);
+                    retriedConnects.inc();
+                } catch (InterruptedException e) {
+                    LOGGER.warn("Sleep interrupted while waiting for reconnectInterval <{}> on <[{}]>:<[{}]>", reconnectInterval, relpAddress, relpPort, e);
```
reconnectInterval is user supplied variable so it must be logged with brackets: <[{}]>

kortemik commented on src/main/java/com/teragrep/aer_01/Output.java:
```diff
@@ -43,152 +43,21 @@
-                }
-            }
+public interface Output extends Consumer<byte[]>, AutoCloseable {
+    final class FakeOutput implements Output {
```
Fakes must exists only within test/ scope, please move the class.
